### ISSUE TITLE
Advancing my "Start of next day" advanced several cards by additional years

### ISSUE THREAD

**Issue by wchristian** [ankidroid/Anki-Android#3801]
I'm not 100% sure if it IS the root cause, but it is the only thing i can think of that would be a smoking gun.

What happened was that, at 4AM local time (with `start of next day` set to 5 hours) i noticed in the statistics that ankidroid apparently thought it was already the next day. (UTC issue? I'm at UTC+1.)

So i decided to set the `start of next day` one hour forward to 6.

Afterwards i noticed that in the statistics the forecast was empty.

I figured it was a display error and went on my merry way.

2 days later i noticed that i wasn't getting any "young" review cards.

I synched to desktop and inspected the data in the card browser, and found that numerous cards were suddenly scheduled for 2018, some with only an interval of a few days.

So, something messed up the data in the sqlite db itself.

Inspecting the data, the relevant cards come up with:

    select * from cards where type = 2 and did = 1403740766368 order by due 

For now i've fixed my own data by applying the following, where due - 858 sets it roughly to ivl - age_in_days.

    update cards set due = due - 858 where type = 2 and did = 1403740766368 and due > 800

It's also notable that "check database" did nothing in this case.

In the next post i'll try to attach a file with backups of the desktop and ankidroid clients respectively before and after the corruption.

**Comment by wchristian** [comment]
Here are the mentioned backup files:

https://dl.dropboxusercontent.com/u/10190786/in_TYOOL2018_anki_said.zip

**Comment by hssm** [comment]
Thanks. We had a report on the forum about a similar problem, so I'm a bit worried about this one now. 

I'll try to find any upstream bugs related to this first in case the problem is with the scheduling code itself.

**Comment by rpgrca** [comment]
That's exactly what happened. On week days I start a new day at 17, but on weekends I change that to 21 (which I switch back to 17 on Mondays). I've done this several times before without problem, this happened with the last or the previous to last beta version. I would also point out (in my case) that cards that weren't seen yet (without due date but instead just a number) weren't rescheduled to 2018.

**Comment by hssm** [comment]
Changing the `Start of next day` setting in AnkiDroid clobbers your database creation time with some bogus data. Doesn't matter when you do it or what value. Don't touch that setting until I get a fix out.

Broken by [this commit](https://github.com/ankidroid/Anki-Android/commit/c4696285146a9254c756743860bcb7069ff5b86b).

**Comment by timrae** [comment]
Oops, was that the `GregorianCalendar()` change? I was trying to fix a lint warning but maybe I didn't comprehend what the old code was actually doing.

**Comment by hssm** [comment]
Yeah that's the one. I'll admit it's not easy to understand. Still trying to figure it out.

**Comment by eginhard** [comment]
You'll need to add the Calendar member variable again and clone from it instead of creating a new one in onSharedPreferenceChanged().

**Comment by hssm** [comment]
I ended up inlining it all without saving and cloning. Not sure if there is some subtlety I'm missing. I'll send a PR in a bit.

**Comment by timrae** [comment]
It looks like github doesn't do the automatic issue closing on feature branches :-/ not sure if it's a bug

**Comment by hssm** [comment]
Yeah I noticed. I thought maybe I typed it in wrong (fixes vs fixed) but I guess it's not my fault after all. Hmm...

**Comment by hssm** [comment]
>once the commit is merged into the default branch.

Okay it's a feature then.

**Comment by timrae** [comment]
Maybe the issues will all be closed when I run the branch sync script then
